Check that the KMS key ID is valid

tomelliff opened the pull request:
KMS key IDs can be given as either a raw key ID, an alias (always prefixed with 'alias/'), the full ARN to the KMS key ID or the full ARN to the KMS key alias.

I'm not 100% sure about the structure of the change so happy for any feedback on moving things around but felt like it was complex enough of a validation to split it out to a separate function. As we're not testing the function directly and instead just testing the `Prepare` method this might not really be valid but open to suggestions on other approaches.

Should help with https://github.com/hashicorp/packer/issues/6599 but probably isn't enough to fully close it.

SwampDragons commented:
I'm happy to have it split into a separate function; I think it helps keep the code clean. I'll give this a quick couple of tests and then merge. Thanks a lot for this!

tomelliff commented:
Just looked at it myself and not sure why I'm using an array instead of a slice in the test cases so going to fix that now before you start to review it.

SwampDragons commented:
Sounds good! I'll go work on something else for a few so you don't feel like you're racing a clock.

tomelliff commented:
Fixed that mistake, apologies for that.

tomelliff commented:
Hmmm... no idea why changing that to a slice from an array is causing the race detector to fail when it wasn't previously :/

SwampDragons commented:
Probably wasn't you. We have a race that's failing intermittently in our build right now and haven't been able to find the root cause. I'm rerunning the build for you.

tomelliff commented:
Ah okay, thanks for that. Looks like it's passing now.

SwampDragons commented on builder/amazon/common/ami_config.go:
```diff
@@ -128,3 +146,23 @@ func (c *AMIConfig) prepareRegions(accessConfig *AccessConfig) (errs []error) {
+
+// See https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CopyImage.html
+func validateKmsKey(kmsKey string) (valid bool) {
+	kmsKeyIdPattern := `[a-zA-Z1-6-]+$`
```
```suggestion
	kmsKeyIdPattern := `[a-zA-Z0-9-]+$`
```

tomelliff replied:
Switched to checking for the hexadecimal digits in the UUID instead of what my sleep deprived brain told me was the regex for hex.

tomelliff commented:
Derp, that was dumb. I took a quick glance, saw that they were hexadecimal and then my brain did something weird there and thought hex was `[a-z1-6]` :/

Fixed that now and added a few extra test cases too.

Also, now that I'm looking at it not on the train home I've realised I'm not actually testing that the keys in `AMIRegionKMSKeyIDs` are being validated. I'll try and get a chance to look at that when I get a chance later on today.

SwampDragons commented:
Happens to the best of us! 😆

tomelliff commented:
Thanks for the super quick feedback and merge!
